vector/src/main/java/im/vector/app/features/location/LocationTracker.kt

```diff
@@ -36,62 +42,90 @@ class LocationTracker @Inject constructor(
     private val locationManager = context.getSystemService<LocationManager>()
 
     interface Callback {
+        /**
+         * Called on every location update.
+         */
         fun onLocationUpdate(locationData: LocationData)
-        fun onLocationProviderIsNotAvailable()
+
+        /**
+         * Called when no location provider is available to request location updates.
+         */
+        fun onNoLocationProviderAvailable()
     }
 
-    private val callbacks = mutableListOf<Callback>()
+    @VisibleForTesting
+    val callbacks = mutableListOf<Callback>()
 
-    private var hasGpsProviderLiveLocation = false
+    @VisibleForTesting
+    var hasLocationFromFusedProvider = false
+
+    @VisibleForTesting
+    var hasLocationFromGPSProvider = false
 
     private var lastLocation: LocationData? = null
 
+    private val debouncer = Debouncer(createBackgroundHandler(BKG_HANDLER_NAME))
+
     @RequiresPermission(anyOf = [Manifest.permission.ACCESS_COARSE_LOCATION, Manifest.permission.ACCESS_FINE_LOCATION])
     fun start() {
-        Timber.d("## LocationTracker. start()")
-        hasGpsProviderLiveLocation = false
+        Timber.d("start()")
 
         if (locationManager == null) {
-            callbacks.forEach { it.onLocationProviderIsNotAvailable() }
-            Timber.v("## LocationTracker. LocationManager is not available")
+            callbacks.forEach { it.onNoLocationProviderAvailable() }
+            Timber.v("LocationManager is not available")
             return
         }
 
-        locationManager.allProviders
-                .takeIf { it.isNotEmpty() }
-                // Take GPS first
-                ?.sortedByDescending { if (it == LocationManager.GPS_PROVIDER) 1 else 0 }
-                ?.forEach { provider ->
-                    Timber.d("## LocationTracker. track location using $provider")
+        val providers = locationManager.allProviders
 
-                    // Send last known location without waiting location updates
-                    locationManager.getLastKnownLocation(provider)?.let { lastKnownLocation ->
+        if (providers.isEmpty()) {
+            callbacks.forEach { it.onNoLocationProviderAvailable() }
+            Timber.v("There is no location provider available")
+        } else {
+            // Take GPS first
+            providers.sortedByDescending { getProviderPriority(it) }
```

Comment:
use method reference?

Reply:
Done, see [af0ae61](https://github.com/vector-im/element-android/pull/5923/commits/af0ae6123b9a7a9a1249cd9a1c672db9268fef0f). Should we prioritize this syntax as a general rule?

Reply:
We might yes!